qiskit/transpiler/passes/optimization/collect_linear_functions.py

```diff
@@ -0,0 +1,84 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2017, 2021.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+
+"""Replace each sequence of CX and SWAP gates by a single LinearFunction gate."""
+
+from qiskit.circuit.library.standard_gates import CXGate, SwapGate
+from qiskit.circuit.library.generalized_gates import LinearFunction
+from qiskit.transpiler.basepasses import TransformationPass
+from qiskit.circuit import QuantumCircuit
+
+
+class CollectLinearFunctions(TransformationPass):
+    """Collect blocks of linear gates (CX and SWAP) and replace them by linear functions."""
+
+    @staticmethod
+    def _is_linear_gate(op):
+        return isinstance(op, (CXGate, SwapGate)) and op.condition is None
```

Comment:
```suggestion
        return op.name in ('cx', 'swap') and op.condition is None
```
This should be faster to check, and we don't generally support sub-classing standard gates (though this isn't yet 100% settled, related to the last point in https://github.com/Qiskit/qiskit-terra/pull/7361#pullrequestreview-825461462 .)